fastfield_codecs/src/compact_space.rs

```diff
@@ -0,0 +1,783 @@
+        blank_collector.stage_blank(delta_and_pos);
+
+        let staged_spaces_sum: u128 = blank_collector.staged_blanks_sum();
+        // +1 for later added null value
+        let amplitude_new_compact_space = amplitude_compact_space - staged_spaces_sum + 1;
+        let amplitude_new_bits = (amplitude_new_compact_space as f64).log2().ceil() as u8;
+        if amplitude_bits == amplitude_new_bits {
+            continue;
+        }
+        let saved_bits = (amplitude_bits - amplitude_new_bits) as usize * values_sorted.len();
+        let cost = blank_collector.num_blanks() * cost_per_blank;
+        if cost >= saved_bits {
+            // Continue here, since although we walk over the deltas by size,
+            // we can potentially save a lot at the last bits, which are smaller deltas
+            //
+            // E.g. if the first range reduces the compact space by 1000 from 2000 to 1000, which
+            // saves 11-10=1 bit and the next range reduces the compact space by 950 to
+            // 50, which saves 10-6=4 bit
+            continue;
+        }
+
+        amplitude_compact_space = amplitude_new_compact_space;
+        amplitude_bits = amplitude_new_bits;
+        for pos in blank_collector
+            .drain()
+            .map(|blank_and_pos| blank_and_pos.pos)
+        {
+            let blank_end = values_sorted[pos] - 1;
+            let blank_start = if pos == 0 {
+                0
+            } else {
+                values_sorted[pos - 1] + 1
+            };
+            compact_space.add_blank(blank_start..=blank_end);
+        }
+    }
+    compact_space.add_blank(max_val..=u128::MAX);
+
+    compact_space.finish()
+}
+
+#[test]
+fn compact_space_test() {
+    let ips = vec![
+        2u128, 4u128, 1000, 1001, 1002, 1003, 1004, 1005, 1008, 1010, 1012, 1260,
+    ];
+    let compact_space = get_compact_space(&ips, 11);
+    assert_eq!(compact_space.null_value, 5);
+    let amplitude = compact_space.amplitude_compact_space();
+    assert_eq!(amplitude, 20);
+    assert_eq!(2, compact_space.to_compact(2).unwrap());
+    assert_eq!(compact_space.to_compact(100).unwrap_err(), 0);
+
+    let mut output = vec![];
+    compact_space.serialize(&mut output).unwrap();
+
+    assert_eq!(
+        compact_space,
+        CompactSpace::deserialize(&mut &output[..]).unwrap()
+    );
+
+    for ip in &ips {
+        let compact = compact_space.to_compact(*ip).unwrap();
+        assert_eq!(compact_space.unpack(compact), *ip);
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+struct CompactSpaceBuilder {
+    covered_space: Vec<RangeInclusive<u128>>,
+}
+
+impl CompactSpaceBuilder {
+    /// Creates a new compact space builder which will initially cover the whole space.
+    fn new() -> Self {
+        Self {
+            covered_space: vec![0..=u128::MAX],
+        }
+    }
+
+    /// Will extend the first range and assign the null value to it.
+    fn assign_and_return_null(&mut self) -> u128 {
```

Comment:
Is there an extra cost 0 to using 0 as the null value?

I suspect it might help the simplicity of the code.

Reply:
If we choose 0 we need to shift all numbers by one, but we can't shift if U128::MAX is occupied.